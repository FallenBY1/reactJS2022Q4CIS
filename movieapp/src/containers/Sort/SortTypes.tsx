export type Sort = {
  [id: string]: string;
  title: string;
  value: string;
};
