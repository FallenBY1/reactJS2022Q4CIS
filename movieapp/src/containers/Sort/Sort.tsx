import { Sort } from './SortTypes';
import { Option } from '../../components/Option/Option';
import { useDispatch } from 'react-redux';
import { receiveMoviesFromApiError, sortMoviesById } from '../../actions/actions';
import { AnyAction } from 'redux';
import { searchParamsToQueryString, updateBrowserUrl } from '../../services/utils';
import { SortCriteria } from '../../services/constants';
import { useTranslation } from 'react-i18next';
import { useNavigate } from 'react-router-dom';

export function Sort(): JSX.Element {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const movieSortCriterias: Array<Sort> = [
    {
      id: '1',
      title: t('popularity'),
      value: SortCriteria.VOTE_COUNT
    },
    {
      id: '2',
      title: t('release_date'),
      value: SortCriteria.RELEASE_DATE
    },
    {
      id: '3',
      title: t('rating'),
      value: SortCriteria.VOTE_AVERAGE
    }
  ];

  function sortMovies(e: any): void {
    const searchParams = { sortBy: e.currentTarget.value };
    navigate('?' + updateBrowserUrl(searchParams), { replace: true });
    fetch(searchParamsToQueryString(searchParams))
      .then((res) => res.json())
      .then((movies) => {
        dispatch(sortMoviesById(movies.data) as AnyAction);
      })
      .catch((error) => dispatch(receiveMoviesFromApiError(error) as AnyAction));
  }

  return (
    <select onChange={sortMovies}>
      {movieSortCriterias.map((element) => (
        <Option key={element.id} title={element.title} value={element.value} />
      ))}
    </select>
  );
}
