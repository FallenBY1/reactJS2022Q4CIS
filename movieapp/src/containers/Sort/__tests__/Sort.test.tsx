import renderer from 'react-test-renderer';
import { describe } from '@jest/globals';
import { Sort } from '../Sort';
import { fireEvent, render } from '@testing-library/react';
import fetchMock from 'jest-fetch-mock';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: any) => str,
      i18n: {
        changeLanguage: () => new Promise(() => {})
      }
    };
  }
}));

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch
}));

jest.mock('react-router', () => ({
  useNavigate: () => mockDispatch,
  navigate: () => mockDispatch
}));

const setup = () => {
  const utils = render(<Sort />);
  const select = utils.getByLabelText('sort');
  return {
    select,
    ...utils
  };
};

beforeEach(() => {
  fetchMock.doMock();
});

describe('Sort container', () => {
  it('should be rendered', () => {
    const sort = renderer.create(<Sort />).toJSON();
    expect(sort).toMatchSnapshot();
  });

  it('should fire change event', function () {
    const { select } = setup();
    const sortMovies = jest.fn();
    fireEvent.change(select);
    expect(sortMovies).toHaveBeenCalled();
  });
});
