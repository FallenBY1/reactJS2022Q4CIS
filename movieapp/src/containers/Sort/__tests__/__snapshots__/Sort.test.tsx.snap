// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort container should be rendered 1`] = `
<select
  aria-label="sort"
  onChange={[Function]}
>
  <option
    value="vote_count"
  >
    popularity
  </option>
  <option
    value="release_date"
  >
    release_date
  </option>
  <option
    value="vote_average"
  >
    rating
  </option>
</select>
`;
